File: 4_Kleisli_Categories/optional.cpp
#include <iostream>
#include <cmath>

// Implemented in C++17, here is a simple implementation of the optional class
template <typename T> class optional {
    bool _isValid;
    T _value;
public:
    optional() : _isValid(false) {}
    optional(T value) : _isValid(true), _value(value) {}
    bool isValid() const { return _isValid; }
    T value() const { return _value; }
};


// Function to calculate the square root safely
optional<double> safe_root(double x) {
    if (x < 0) {
        return optional<double>{}; // Return an empty optional if the input is negative
    }
    return optional<double>{std::sqrt(x)}; // Return the square root if the input is non-negative
}

int main() {
    double values[] = {4.0, -1.0, 9.0, 0.0, 16.0};

    for (double value : values) {
        optional<double> result = safe_root(value);

        if (result.isValid()) {
            std::cout << "The square root of " << value << " is " << result.value() << std::endl;
        } else {
            std::cout << "Cannot calculate the square root of " << value << std::endl;
        }
    }

    return 0;
}

File: 4_Kleisli_Categories/optional.hs
-- Define the Optional type
data Optional a = None | Some a
    deriving (Show, Eq)

-- Identity morphism for Optional
return :: a -> Optional a
return = Some

-- Function to compose two functions that return Optional values
(>=>) :: (a -> Optional b) -> (b -> Optional c) -> (a -> Optional c)
f >=> g = \x -> case f x of
    None   -> None
    Some y -> g y

-- Function to safely calculate the square root
safeRoot :: (Ord a, Floating a) => a -> Optional a
safeRoot x
    | x < 0     = None
    | otherwise = Some (sqrt x)

-- Function to safely calculate the inverse
safeInverse :: (Eq a, Fractional a) => a -> Optional a
safeInverse x
    | x == 0    = None
    | otherwise = Some (1 / x)

-- Example usage
main :: IO ()
main = do
    let values = [4.0, -1.0, 9.0, 0.0, 16.0]
    let safeRootInverse = safeRoot >=> safeInverse
    mapM_ (printResult safeRootInverse) values

-- Helper function to print the result
printResult :: (Show a, Show b) => (a -> Optional b) -> a -> IO ()
printResult f x = case f x of
    None   -> putStrLn $ "Cannot calculate the result for " ++ show x
    Some y -> putStrLn $ "The result for " ++ show x ++ " is " ++ show y